drivers/platform/maxim/common/maxim_dma.c

```diff
@@ -0,0 +1,289 @@
+ * modification, are permitted provided that the following conditions are met:
+ *  - Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *  - Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *  - Neither the name of Analog Devices, Inc. nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <stdlib.h>
+#include "no_os_error.h"
+#include "no_os_alloc.h"
+#include "no_os_util.h"
+#include "no_os_irq.h"
+#include "maxim_irq.h"
+#include "maxim_dma.h"
+#include "mxc_sys.h"
+#include "dma.h"
+
+#define MAX_DMA_GCR_RST_POS	0
+
+/**
+* @brief There is only one DMA controller
+*/
+static struct no_os_dma_desc *dma_descriptor;
+
+/**
+ * @brief Initialize a DMA controller.
+ * @param desc - Descriptor to be initialized.
+ * @param param - Initialization parameter for the decriptor.
+ * @return 0 in case of success
+ * 	   -ENOMEM if there is not enough free memory
+ */
+static int maxim_dma_init(struct no_os_dma_desc **desc,
+			  struct no_os_dma_init_param *param)
+{
+	struct no_os_dma_desc *descriptor;
+	uint32_t i;
+	int ret;
+
+	struct no_os_irq_init_param irq_param = {
+		.irq_ctrl_id = 0,
+		.platform_ops = &max_irq_ops,
+		.extra = NULL
+	};
+
+	if (dma_descriptor) {
+		/*
+		 * There is only one DMA controller and it's already
+		 * initialized.
+		 */
+		*desc = dma_descriptor;
+		return 0;
```

Comment:
In the Maxim platform specific case, there is really nothing to update. The id for a specific interface should remain constant. The same goes for the number of channels. However, I'm not entirely sure how should we handle other cases where this reference counting mechanism would be used (say for example a driver which would allow making updates). Because in that case an update would result in changing other references already held.